### PR TITLE
[Summer of code] Record haconnection's lastCaughtUpTimeMs in haService's map

### PR DESCRIPTION
**Pull request by hzh0425**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

tracking issue: https://github.com/apache/rocketmq/issues/4330

## Brief changelog

Record haconnection's lastCaughtUpTimeMs in haService's map

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java [hunk 1]
```diff
@@ -162,15 +165,13 @@ public Set<String> maybeShrinkInSyncStateSet() {
...
             if (currentSyncStateSet.contains(slaveAddress)) {
-                if (connection.getSlaveAckOffset() < 0 || this.defaultMessageStore.getMaxPhyOffset() == connection.getSlaveAckOffset()) {
-                    continue;
-                }
-                if ((System.currentTimeMillis() - connection.getLastCatchUpTimeMs()) > haMaxTimeSlaveNotCatchup) {
+                final Long lastCaughtUpTimeMs = this.connectionCaughtUpTimeTable.get(slaveAddress);
+                if ((System.currentTimeMillis() - lastCaughtUpTimeMs) > haMaxTimeSlaveNotCatchup) {
                     newSyncStateSet.remove(slaveAddress);
+                    this.connectionCaughtUpTimeTable.remove(slaveAddress);
                 }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java:175]: 这里存在这样一种情况：connectionCaughtUpTimeTable和newSyncStateSet把slave移除了，但newSyncStateSet上报没有成功，所以syncStateSet并没有修改成功。但下次shrink的时候由于connectionCaughtUpTimeTable已经移除了，syncStateSet不会移除了。

**Comment by RongtongJin** [comment]
Failed tests: 
  AutoSwitchRoleIntegrationTest.testCheckSyncStateSet:117 expected:<1> but was:<2>

File: store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java (hunk 1 above)
Comment (hzh0425) [reply store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java:175]: 那 connectionCaughtUpTimeTable 等 syncStateSet 上报成功后再移除么

Comment (RongtongJin) [reply store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java:175]: > 那 connectionCaughtUpTimeTable 等 syncStateSet 上报成功后再移除么

可以的。

**Pull request merged**